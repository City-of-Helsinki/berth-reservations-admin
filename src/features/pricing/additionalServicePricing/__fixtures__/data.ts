import { AdditionalServicePricingProps } from '../AdditionalServicePricing';
import {
  ProductServiceType,
  PriceUnits,
  PeriodType,
  AdditionalProductType,
  AdditionalProductTaxEnum,
} from '../../../../@types/__generated__/globalTypes';

export const data: AdditionalServicePricingProps['data'] = {
  __typename: 'AdditionalProductNodeConnection',
  edges: [
    {
      __typename: 'AdditionalProductNodeEdge',
      node: {
        __typename: 'AdditionalProductNode',
        id: 'a8400b59-534c-4934-b643-083a4273ca1a',
        service: ProductServiceType.DINGHY_PLACE,
        priceValue: 10,
        priceUnit: PriceUnits.AMOUNT,
        period: PeriodType.SEASON,
        productType: AdditionalProductType.OPTIONAL_SERVICE,
        taxPercentage: AdditionalProductTaxEnum.TAX_24_00,
      },
    },
  ],
};
