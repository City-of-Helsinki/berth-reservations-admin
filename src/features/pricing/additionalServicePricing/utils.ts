import { AdditionalServicePricingProps, AdditionalService } from './AdditionalServicePricing';

export const getAdditionalServiceData = (data: AdditionalServicePricingProps['data']): AdditionalService[] => {
  if (!data) return [];

  return data.edges.reduce<AdditionalService[]>((acc, edge) => {
    if (!edge?.node) return acc;

    const berthPrice = {
      id: edge.node.id,
      service: edge.node.service,
      price: edge.node.priceValue,
      tax: edge.node.taxPercentage,
      period: edge.node.period,
    };

    return [...acc, berthPrice];
  }, []);
};
