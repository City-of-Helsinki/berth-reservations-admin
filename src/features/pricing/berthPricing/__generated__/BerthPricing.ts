/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { PriceUnits } from "./../../../../@types/__generated__/globalTypes";

// ====================================================
// GraphQL fragment: BerthPricing
// ====================================================

export interface BerthPricing_edges_node_defaultProduct {
  __typename: "BerthProductNode";
  priceUnit: PriceUnits;
  priceValue: any;
}

export interface BerthPricing_edges_node {
  __typename: "BerthPriceGroupNode";
  id: string;
  name: string;
  defaultProduct: BerthPricing_edges_node_defaultProduct | null;
}

export interface BerthPricing_edges {
  __typename: "BerthPriceGroupNodeEdge";
  node: BerthPricing_edges_node | null;
}

export interface BerthPricing {
  __typename: "BerthPriceGroupNodeConnection";
  edges: (BerthPricing_edges | null)[];
}
