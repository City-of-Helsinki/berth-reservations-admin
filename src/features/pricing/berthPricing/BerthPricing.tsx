import React, { useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Row } from 'react-table';
import { useMutation } from '@apollo/react-hooks';

import Card from '../../../common/card/Card';
import CardHeader from '../../../common/cardHeader/CardHeader';
import CardBody from '../../../common/cardBody/CardBody';
import Table, { Column, COLUMN_WIDTH } from '../../../common/table/Table';
import Section from '../../../common/section/Section';
import Text from '../../../common/text/Text';
import { formatPrice } from '../../../common/utils/format';
import EditForm, { EDIT_FORM_TYPE } from '../editModal/EditForm';
import { BerthPricing as BerthPricingData } from './__generated__/BerthPricing';
import { getBerthsData } from './utils';
import { PeriodType } from '../../../@types/__generated__/globalTypes';
import { getPeriodTKey } from '../../../common/utils/translations';
import EditModal from '../editModal/EditModal';
import { UPDATE_BERTH_PRICE_MUTATION, CREATE_BERTH_PRODUCT_MUTATION } from './mutations';
import {
  UPDATE_BERTH_PRICE,
  UPDATE_BERTH_PRICEVariables as UPDATE_BERTH_PRICE_VARS,
} from './__generated__/UPDATE_BERTH_PRICE';
import {
  CREATE_BERTH_PRODUCT,
  CREATE_BERTH_PRODUCTVariables as CREATE_BERTH_PRODUCT_VARS,
} from './__generated__/CREATE_BERTH_PRODUCT';

export interface BerthPrice {
  id: string;
  productId: string | undefined;
  name: string;
  privateCustomer: number | undefined;
  company: number | undefined;
  period: PeriodType;
}

export interface BerthPricingProps {
  className?: string;
  data: BerthPricingData | undefined | null;
  loading: boolean;
}

const BerthPricing = ({ className, data, loading }: BerthPricingProps) => {
  const { t, i18n } = useTranslation();
  const [editRowValues, setEditRowValues] = useState<BerthPrice>();
  const [updateBerthPrice] = useMutation<UPDATE_BERTH_PRICE, UPDATE_BERTH_PRICE_VARS>(UPDATE_BERTH_PRICE_MUTATION);
  const [createBerthPrice] = useMutation<CREATE_BERTH_PRODUCT, CREATE_BERTH_PRODUCT_VARS>(
    CREATE_BERTH_PRODUCT_MUTATION
  );

  const harborCols: Column<BerthPrice>[] = [
    {
      Header: t('pricing.berths.width') || '',
      accessor: 'name',
    },
    {
      Header: t('pricing.berths.privateCustomer') || '',
      accessor: 'privateCustomer',
      Cell: ({ cell: { value } }) => (value ? formatPrice(value, i18n.language) : '-'),
    },
    {
      Header: t('pricing.berths.company') || '',
      accessor: 'company',
      Cell: ({ cell: { value } }) => (value ? formatPrice(value, i18n.language) : '-'),
    },
    {
      Header: t('pricing.berths.period') || '',
      accessor: 'period',
      Cell: ({ cell }) => t(getPeriodTKey(cell.value)),
    },
    {
      id: 'edit',
      Header: t('common.edit') || '',
      sortType: 'none',
      width: COLUMN_WIDTH.S,
      Cell: ({ row }: { row: Row<BerthPrice> }) => (
        <button onClick={() => setEditRowValues(row.original)}>
          <Text color="brand">{t('common.edit')}</Text>
        </button>
      ),
    },
  ];

  const handleSubmit = async ({
    id: priceGroupId,
    productId: defaultProductId,
    privateCustomer: priceValue,
  }: BerthPrice): Promise<void> => {
    if (defaultProductId) {
      await updateBerthPrice({
        variables: { input: { id: defaultProductId, priceValue } },
      });
    } else {
      await createBerthPrice({
        variables: { input: { priceGroupId, priceValue } },
      });
    }

    setEditRowValues(undefined);
  };

  const handleClose = () => setEditRowValues(undefined);

  return (
    <>
      <Card className={className}>
        <CardHeader title={t('pricing.berths.title')} />
        <CardBody>
          <Section>{t('pricing.berths.description')}</Section>
          <Table
            columns={harborCols}
            data={getBerthsData(data)}
            loading={loading}
            theme="basic"
            renderEmptyStateRow={() => t('common.notification.noData.description')}
          />
        </CardBody>
      </Card>
      <EditModal isOpen={!!editRowValues} closeModal={handleClose}>
        {editRowValues && (
          <EditForm
            closeModal={handleClose}
            formType={EDIT_FORM_TYPE.BERTHS}
            initialValues={editRowValues}
            onSubmit={handleSubmit}
          />
        )}
      </EditModal>
    </>
  );
};

export default BerthPricing;
