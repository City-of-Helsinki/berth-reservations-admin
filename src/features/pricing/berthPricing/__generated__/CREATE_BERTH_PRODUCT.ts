/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { CreateBerthProductMutationInput, PriceUnits } from "./../../../../@types/__generated__/globalTypes";

// ====================================================
// GraphQL mutation operation: CREATE_BERTH_PRODUCT
// ====================================================

export interface CREATE_BERTH_PRODUCT_createBerthProduct_berthProduct {
  __typename: "BerthProductNode";
  id: string;
  priceValue: any;
  priceUnit: PriceUnits;
}

export interface CREATE_BERTH_PRODUCT_createBerthProduct {
  __typename: "CreateBerthProductMutationPayload";
  berthProduct: CREATE_BERTH_PRODUCT_createBerthProduct_berthProduct | null;
}

export interface CREATE_BERTH_PRODUCT {
  createBerthProduct: CREATE_BERTH_PRODUCT_createBerthProduct | null;
}

export interface CREATE_BERTH_PRODUCTVariables {
  input: CreateBerthProductMutationInput;
}
