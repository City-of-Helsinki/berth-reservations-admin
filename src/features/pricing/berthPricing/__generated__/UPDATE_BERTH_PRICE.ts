/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { UpdateBerthProductMutationInput, PriceUnits } from "./../../../../@types/__generated__/globalTypes";

// ====================================================
// GraphQL mutation operation: UPDATE_BERTH_PRICE
// ====================================================

export interface UPDATE_BERTH_PRICE_updateBerthProduct_berthProduct {
  __typename: "BerthProductNode";
  id: string;
  priceValue: any;
  priceUnit: PriceUnits;
}

export interface UPDATE_BERTH_PRICE_updateBerthProduct {
  __typename: "UpdateBerthProductMutationPayload";
  berthProduct: UPDATE_BERTH_PRICE_updateBerthProduct_berthProduct | null;
}

export interface UPDATE_BERTH_PRICE {
  updateBerthProduct: UPDATE_BERTH_PRICE_updateBerthProduct | null;
}

export interface UPDATE_BERTH_PRICEVariables {
  input: UpdateBerthProductMutationInput;
}
