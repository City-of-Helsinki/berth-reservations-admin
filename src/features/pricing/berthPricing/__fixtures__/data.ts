import { PlaceProductTaxEnum, PriceUnits, PricingCategory } from '../../../../@types/__generated__/globalTypes';
import { BerthPricingProps } from '../BerthPricing';

export const data: BerthPricingProps['data'] = {
  __typename: 'BerthProductNodeConnection',
  edges: [
    {
      __typename: 'BerthProductNodeEdge',
      node: {
        __typename: 'BerthProductNode',
        id: '68ac7db5-b397-48f6-af19-99c35d645cb6',
        minWidth: 3,
        maxWidth: 7,
        tier1Price: 50,
        tier2Price: 75,
        tier3Price: 100,
        priceUnit: PriceUnits.AMOUNT,
        taxPercentage: PlaceProductTaxEnum.TAX_24_00,
        pricingCategory: PricingCategory.DEFAULT,
      },
    },
    {
      __typename: 'BerthProductNodeEdge',
      node: {
        __typename: 'BerthProductNode',
        id: '82c3ba8c-f409-4f54-ad8f-12ace3f5df50',
        minWidth: 6,
        maxWidth: 14,
        tier1Price: 150,
        tier2Price: 175,
        tier3Price: 200,
        priceUnit: PriceUnits.AMOUNT,
        taxPercentage: PlaceProductTaxEnum.TAX_24_00,
        pricingCategory: PricingCategory.DEFAULT,
      },
    },
    {
      __typename: 'BerthProductNodeEdge',
      node: {
        __typename: 'BerthProductNode',
        id: '27a112bd-734c-461c-86d1-ba78d7351a3f',
        minWidth: 3,
        maxWidth: 7,
        tier1Price: 40,
        tier2Price: 60,
        tier3Price: 80,
        priceUnit: PriceUnits.AMOUNT,
        taxPercentage: PlaceProductTaxEnum.TAX_24_00,
        pricingCategory: PricingCategory.VASIKKASAARI,
      },
    },
  ],
};
