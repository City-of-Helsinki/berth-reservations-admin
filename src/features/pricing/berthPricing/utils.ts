import i18next from 'i18next';
import { BerthPricing as BerthPricingData } from './__generated__/BerthPricing';
import { BerthPrice } from './BerthPricing';
import { PriceTier, PricingCategory } from '../../../@types/__generated__/globalTypes';
import { formatDimension } from '../../../common/utils/format';

type BerthPriceData = Record<PricingCategory, BerthPrice[]>;

export const getBerthsData = (data: BerthPricingData | null | undefined): BerthPriceData => {
  const initialData: BerthPriceData = {
    [PricingCategory.DEFAULT]: [],
    [PricingCategory.DINGHY]: [],
    [PricingCategory.TRAILER]: [],
    [PricingCategory.VASIKKASAARI]: [],
  };
  if (!data) return initialData;

  return data.edges.reduce<BerthPriceData>((acc, edge) => {
    if (!edge?.node) return acc;

    const berthPrice = {
      id: edge.node.id,
      productId: edge.node.id,
      name: `${formatDimension(edge.node.minWidth, i18next.language)} - ${formatDimension(
        edge.node.maxWidth,
        i18next.language
      )}`,
      [PriceTier.TIER_1]: edge.node.tier1Price,
      [PriceTier.TIER_2]: edge.node.tier2Price,
      [PriceTier.TIER_3]: edge.node.tier3Price,
    };

    const pricingCategory = edge.node.pricingCategory;

    return { ...acc, [pricingCategory]: [...acc[pricingCategory], berthPrice] };
  }, initialData);
};
