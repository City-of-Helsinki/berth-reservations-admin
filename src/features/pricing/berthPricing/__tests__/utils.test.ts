import { getBerthsData } from '../utils';
import { BerthPricing } from '../__generated__/BerthPricing';
import { data } from '../__fixtures__/data';

describe('utils', () => {
  describe('getBerthsData', () => {
    it('should return an array of BerthPrice objects', () => {
      expect(getBerthsData(data)).toMatchSnapshot();
    });

    it("company's fee should be twice as much as the private customer's", () => {
      const berthPrices = getBerthsData(data as BerthPricing);

      expect(berthPrices[0].company).toBe((berthPrices[0].privateCustomer as number) * 2);
    });

    it('should return an empty array when the provided data is undefined or null', () => {
      expect(getBerthsData(undefined)).toEqual([]);
      expect(getBerthsData(null)).toEqual([]);
    });

    it('should remove all potential null/undefined nodes', () => {
      const data = { __typename: '', edges: [undefined, null] };

      expect(getBerthsData(data as BerthPricing)).toEqual([]);
    });
  });
});
