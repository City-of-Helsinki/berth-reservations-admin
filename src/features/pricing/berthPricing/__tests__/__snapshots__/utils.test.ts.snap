// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils getBerthsData should return an array of BerthPrice objects 1`] = `
Array [
  Object {
    "company": 20.4,
    "id": "68ac7db5-b397-48f6-af19-99c35d645cb6",
    "name": "sed sed debitis",
    "period": "SEASON",
    "privateCustomer": 10.2,
  },
]
`;
